New module vip-publication out of the existing publication in vip-core

camarasu opened the pull request:
Better modularity and the possibility to associate a VIPApplication to a publication (see Issue #220)

Changes/Improvements : 
- separate VIPPublication module
- Added a VIPApplication field in the Publication form and to VIPPublication DB (issue #220)
- VIPPublication DB is now created automatically if it does not exists
- VIPAuthor and VIPApplication are secondary keys, which allows automatic update (removal of the updateAuthor method)
- update of getApplications() to work for non-admin users also, by returning only the list of applications they have access to

axlbonnet commented on vip-publication/src/main/java/fr/insalyon/creatis/vip/publication/client/view/Data.java:
```diff
@@ -29,15 +29,12 @@
  * knowledge of the CeCILL-B license and that you accept its terms.
  */
-package fr.insalyon.creatis.vip.core.client.view.user.publication;
+package fr.insalyon.creatis.vip.publication.client.view;
```
Is it not necessary to add vipApplication in this class ? I'm lost with all those gwt grid data holder classes.

camarasu replied:
Done

axlbonnet reviewed (changes requested):
Good job to refactor this publication module ! Some small remarks about the gwt/java code.

Obviously the remaining job concerns database maintenance, in 2 steps : this PR and at release time.
I think we must, with this PR :
- rename the table to the original `VipPublication` in this PR code
- update the engine of the current `VipPublication` (myisam -> innodb)
- prepare the SQL requests to run on release to add the new column and the new constraints (to write in the `database_update.sql` at repository root)
- verify all the current publication vipAuthor actually exist to avoid a constraint issue when we will add the foreign key. 

At release time, we will only have to run the few SQL requests.

camarasu commented on vip-application/src/main/java/fr/insalyon/creatis/vip/application/server/rpc/ApplicationServiceImpl.java:
```diff
@@ -225,8 +226,13 @@ public String publishVersion(String applicationName, String version) throws Appl
+            List<String> classNames = new ArrayList<>();
+            for (AppClass c : classes) {
+                classNames.add(c.getName());
+            }
```
Nice, thanks for the suggestion

camarasu commented on vip-publication/src/main/java/fr/insalyon/creatis/vip/publication/client/view/EditPublicationLayout.java:
```diff
@@ -152,6 +161,7 @@ public void setPublication(String id, String title, String type, String nameType
             this.publicationDate.setValue("");
             this.doiField.setValue("");
             this.authorsField.setValue("");
+            this.vipApplication.setValue(vipApplication);
```
Indeed.
I created another ticket to handle compulsory values

camarasu commented:
As discussed, there is no need to update VipPublication, neither to change myisam -> innodb, nor to add the new column and the new constraints, because I 
- created a new table (the code for the table creation is integrated in the PR)
- imported data from the old one
- renamed the new table with the old name (current VIP version works fine with the new table which has one extra column)

I also renamed the table to the original VipPublication in this PR code
So there should be no change required when getting the next release in production

axlbonnet commented:
@camarasu  Doing tests, I found there is an issue with the new `PublicationParser testLastUpdatePublication` call.

The `callback` is `null` and I guess that GWT does not support its `final` field constuction.
